icu4c/source/common/unistr.cpp

```diff
@@ -1543,6 +1552,15 @@ UnicodeString::doAppend(const UChar *srcChars, int32_t srcStart, int32_t srcLeng
     }
   }
 
+  // Check for append onto ourself
```

Comment:
Mostly same comments as above.
There is no oldArray to pull up, but I would create an oldArray to amortize the getArrayStart().